src/dataplane-protocol/src/smartstream.rs

```diff
@@ -0,0 +1,135 @@
+            write!(
+                f,
+                "{}\n\n\
+                SmartStream Info: \n    \
+                Type: {}\n    \
+                Offset: {}\n    \
+                Key: {}\n    \
+                Value: {}",
+                self.hint, self.kind, self.offset, key, value,
+            )
+        }
+    }
+
+    fn display_record_data(record: &RecordData) -> String {
+        match std::str::from_utf8(record.as_ref()) {
+            Ok(s) => s.to_string(),
+            _ => format!("Binary: {} bytes", record.as_ref().len()),
+        }
+    }
+
+    #[derive(Debug, Default, Clone, PartialEq, Encoder, Decoder)]
+    pub struct SmartStreamRuntimeErrorBuilder {
+        /// Error hint: meant for users, not for code
+        hint: String,
+        /// The type of SmartStream that had a runtime error
+        kind: SmartStreamType,
+        /// The offset delta of the Record that had a runtime error
+        // Offset delta is relative to the beginning of a batch
+        offset_delta: Offset,
+        /// The absolute offset, after we calculate it
+        offset: Option<Offset>,
+        /// A preview of the Record key that caused this error
+        record_key: Option<RecordData>,
+        /// A preview of the Record value that caused this error
+        record_value: Option<RecordData>,
+    }
+
+    impl SmartStreamRuntimeErrorBuilder {
+        pub fn new(err: eyre::Error, offset_delta: Offset, kind: SmartStreamType) -> Self {
+            Self {
+                hint: format!("{:?}", err),
+                kind,
+                offset_delta,
+                offset: None,
+                record_key: None,
+                record_value: None,
+            }
+        }
+
+        /// Apply a base offset to this error to derive the absolute offset
+        pub fn base_offset(&mut self, base: Offset) -> &mut Self {
+            self.offset = Some(base + self.offset_delta);
+            self
+        }
+
+        /// Generate a preview of the Record that failed
+        pub fn record(&mut self, record: &Record) -> &mut Self {
+            self.record_key = record.key.clone();
+            self.record_value = Some(record.value.clone());
+            self
+        }
+
+        /// Produce the SmartStreamRuntimeError
+        pub fn build(&self) -> SmartStreamRuntimeError {
+            SmartStreamRuntimeError {
+                offset: self.offset.clone().unwrap(),
+                kind: self.kind.clone(),
+                record_key: self.record_key.clone(),
+                record_value: self.record_value.clone().unwrap(),
+                hint: self.hint.clone(),
+            }
+        }
+    }
+
+    /// A type used to return processed records and/or an error from a SmartStream
+    #[derive(Debug, Default, Encoder, Decoder)]
+    pub struct SmartStreamOutput {
+        pub successes: Vec<Record>,
+        pub error: Option<SmartStreamRuntimeErrorBuilder>,
```

Comment:
shouldn't this be SmartStreamRunTimeError?

Reply:
The `SmartStreamOutput` type is the top-level data object that gets serialized from WASM and transmitted to SPU. At that point, we only have partial data of what we want to collect for the runtime error, specifically the `hint` and the `offset_delta`, but the other pieces of data come from the SPU itself. So we use the builder here because it collects half of the data in WASM and half in the SPU, and then it gets converted to the proper `SmartStreamRuntimeError` right before sending the response to the consumer

Reply:
So this shouldn't belong in here since this is only relevant internally to SPU-WASM.  This create should be contain that protocol between client and cluster (SPU).

Reply:
Where do you think it should go instead? I put it here because `dataplane-protocol` is the common crate that is imported by both the WASM application code and the SPU, so the types belong here so that they are accessible to both. I can't move this just into the SPU because then the WASM couldn't import these types, and vice-versa, I can't move this type into the `fluvio_smartstream` crate because the SPU couldn't import them there.

This seems to be the only place that works, unless we create yet another crate and I really don't want to do that.

Reply:
What about smartstream crate?

Reply:
The smartstream crate is not imported by the SPU, it is only imported by the WASM module itself

Reply:
It is also not something that we would _want_ to import to the SPU, because it exports low-level symbols that are only relevant to the WASM, such as `copy_records` and either `filter` or `map` raw "extern C" functions, which would make no sense to have in the SPU

Reply:
> Probably need new smartstream specific protocol crate

Can we please tackle this in another issue in the future? I am trying so hard to get this V1 shipped